src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -671,17 +677,20 @@ private void ForceAuthentication(bool receiveFirst, byte[] buffer, AsyncProtocol
             // This will tell that we don't know the framing yet (what SSL version is)
             _Framing = Framing.Unknown;
 
+            // Throw if cancellation requested.
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Pretty much any of these places where you've got a ThrowIfCancellationRequested and it's not an async method is likely not going to do the right thing, as the exception will be thrown out synchronously rather than propagate out through the returned Task.

Reply:
Also, do we need all of these polling checks for cancellation?  Seems like there should be a single check up front and then it should just be passed to all of the stream operations.

Reply:
That assumes the underlying stream fully implements cancellation.

Reply:
> That assumes the underlying stream fully implements cancellation.

No, it doesn't.  It assumes the underlying stream at least does a single check for cancellation, which most every implementation does... even the base stream's implementation layered on Begin/End or the sync methods does so.